src/libraries/Common/tests/System/Net/Http/Http3LoopbackConnection.cs

```diff
@@ -0,0 +1,126 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.Net.Quic;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace System.Net.Test.Common
+{
+
+    public sealed class Http3LoopbackConnection : GenericLoopbackConnection
+    {
+        private readonly QuicConnection _connection;
+        private readonly Dictionary<int, Http3LoopbackStream> _openStreams = new Dictionary<int, Http3LoopbackStream>();
+        private Http3LoopbackStream _currentStream;
+
+        public Http3LoopbackConnection(QuicConnection connection)
+        {
+            _connection = connection;
+        }
+
+        public override void Dispose()
+        {
+            foreach (Http3LoopbackStream stream in _openStreams.Values)
+            {
+                stream.Dispose();
+            }
+
+            _connection.Dispose();
+        }
+
+        public Http3LoopbackStream OpenUnidirectionalStream()
+        {
+            return new Http3LoopbackStream(_connection.OpenUnidirectionalStream());
+        }
+
+        public Http3LoopbackStream OpenBidirectionalStream()
+        {
+            return new Http3LoopbackStream(_connection.OpenBidirectionalStream());
+        }
+
+        public static int GetRequestId(QuicStream stream)
+        {
+            Debug.Assert(stream.CanRead && stream.CanWrite, "Stream must be a request stream.");
+            return checked((int)stream.StreamId + 1);
+        }
+
+        public Http3LoopbackStream GetOpenRequest(int requestId = 0)
+        {
+            return requestId == 0 ? _currentStream : _openStreams[requestId - 1];
+        }
+
+        public async Task<Http3LoopbackStream> AcceptStreamAsync()
+        {
+            QuicStream quicStream = await _connection.AcceptStreamAsync().ConfigureAwait(false);
+            var stream = new Http3LoopbackStream(quicStream);
+
+            _openStreams.Add(checked((int)quicStream.StreamId), stream);
+            _currentStream = stream;
+
+            return stream;
+        }
+
+        public override async Task<byte[]> ReadRequestBodyAsync()
+        {
+            return await _currentStream.ReadRequestBodyAsync().ConfigureAwait(false);
```

Comment:
Feels weird to have ReadRequestBodyAsync as part of the connection. Personally I'm not a fan of tracking _currentStream at all.

Reply:
It's how our generic tests work, which are made for HTTP/1 to assume one request at a time.